src/Donations/Routes/RegisterDonationRoutes.php

```
@@ -0,0 +1,132 @@
+                        return $this->donationRequestController->getDonations($request);
+                    },
+                    'permission_callback' => '__return_true',
+                ],
+                'args' => [
+                    'page' => [
+                        'type' => 'integer',
+                        'default' => 1,
+                        'minimum' => 1,
+                    ],
+                    'per_page' => [
+                        'type' => 'integer',
+                        'default' => 30,
+                        'minimum' => 1,
+                        'maximum' => 100,
+                    ],
+                    'sort' => [
+                        'type' => 'string',
+                        'default' => 'id',
+                        'enum' => [
+                            'id',
+                            'createdAt',
+                            'updatedAt',
+                            'status',
+                            'amount',
+                            'feeAmountRecovered',
+                            'donorId',
+                            'firstName',
+                            'lastName',
+                        ],
+                    ],
+                    'direction' => [
+                        'type' => 'string',
+                        'default' => 'DESC',
+                        'enum' => ['ASC', 'DESC'],
+                    ],
+                    'mode' => [
+                        'type' => 'string',
+                        'default' => 'live',
+                        'enum' => ['live', 'test'],
+                    ],
+                    'campaignId' => [
+                        'type' => 'integer',
+                        'default' => 0,
+                    ],
+                    'includeAnonymousDonations' => [
+                        'type' => 'boolean',
+                        'default' => false,
+                    ],
```

@glaubersilva Do we have any restrictions around this parameter? If our goal is to protect anonymous donations, then we may want to include some permission parameters around this. Presently, this is a fully public API, and these parameters are discoverable, so it's not hard for someone to flip the switch on this. There are a few contexts, here:
1. Donor wall
2. Admin list tables (donations list and campaign donations)
3. Campaign donations
4. 3rd Party usage

I'm guessing this will result in 3 forms of output:
1. Anonymous donations are included and donor info revealed (admin-side)
2. Anonymous donations are included but donor information is redacted (donor wall)
3. Anonymous donations are prohibited

Now, we may actually be fine with anyone in the world being able to query anonymous donations so long as the donor information is redacted (dropping option 3). But we'll still need a way to grab all information for the admin side with proper authorization.

Note: this is true of both the collection endpoint and single donation resource endpoint.

cc: @jonwaldstein

---

Exactly. It should throw a 403 if the authorization is invalid for a specific parameter.

---

We already are blocking access to sensitive data for non-admin users:

```
$sensitiveData = [
    'donorIp',
    'email',
    'phone',
    'billingAddress',
 ];
```

If I understand it well, we also should block access to the "anonymous data" when we are dealing with anonymous donations.

So, I refactored the `escDonation()` method and implemented a new logic there to also remove these properties:

```
if ($donation->anonymous) {
    $anonymousData = [
        'donorId',
        'honorific',
        'firstName',
        'lastName',
        'company',
    ];

    $sensitiveData = array_merge($sensitiveData, $anonymousData);
}
```

I considered here "anonymous data" any property from the donation model that can be used to identify a donor, which makes me wonder if we should remove the "comment" property as well. Currently, I kept it on the result, so would like to know your thoughts on it... Should we keep it or not? I'm asking because depending on the content of the comment is possible to identify the donor (I guess).

Also, I wrote more unit tests to check if the access to these "anonymous data" is being blocked by default and allowed for admin users, see: 7def493999265a45cb5d79574386fcdeadaf1bf9 and 238fd0a69a59f9bb86b6d7c936e69dcb1e4f8f99

---

One more thing to consider here is that right now we are removing the "anonymous data" from the results. I was wondering if we should replace the content instead of removing it.

For example:

```
$newValue [
    'donorId' => 'anonymous',
    'honorific' => 'anonymous',
    'firstName' => 'anonymous',
    'lastName' => 'anonymous',
    'company' => 'anonymous',
];
```

---

Thanks @glaubersilva!

I think your'e on the right track.  I would say donation comment can be public but written by _anonymous_.  

Regarding preserving result keys, I'm leaning toward just removing results if not authenticated as I would prefer only receiving real values for my results rather than auto-generated results based on state if that make sense lol

---

I'm thinking that it may be useful to turn `anonymousDonations` into an enum:
* `exclude` — (default) simply excludes the donations from the collection
* `redact` — includes the donations and redacts the sensitive information, replacing with "anonymous"
* `include` — includes the donations with full details (requires authorization)

For `include`, if authorization fails we simply 403 and call it a day, as the client is trying to do something beyond their pay grade.

And I agree, Jon, that the donation comment is public. It's presented to the donor as something public when they chose to write a comment.

What do you guys think?

---

So this is a RESTful design issue, @glaubersilva . You want to think of REST endpoints as similar to a pure function. That is, you get what you expect based on your parameters. Having the shape of the data change based on conditional things such as authorization means that the result is stateful — which is a RESTful no-no. The result should be predictable by the user.

The reason why the enum is a useful approach is that it gives the client explicit control over the shape of the response. The authorization, required for the `include` value, is explicit but requires authorization to work. So either it returns exactly what the client expects or does a 403 because they're not allowed to do that.

So it's not about having a use case where the admin may want the data redacted or not; it's about creating a system wherein the client can explicitly determine what they want so we don't have to worry about the scenario (i.e. state) they're working from. That said, if I'm logged into a site and viewing the donor wall, it should be redacted even if I'm logged in as an admin — otherwise, the inconsistent behavior becomes confusing. If the _client_ wants this behavior, then it can adjust the query parameter based on its own conditions.

Hope this helps!

---

Resolved in 5bd2cc90cd8b99bf22fc531d11bdab4d3524841b 95acf68edd37cf3cee52d9aa6f58e988a13a7e22 06c849945a5d83d494ce824490fb727cf46e427a 8ff58df661b802dfb3f603dbf124e823b3be0624